doc/specs/#5000 - Process Model 2.0/#4472 - Windows Terminal Session Management.md

```diff
@@ -0,0 +1,465 @@
+Spec]. As a quick summary:
+
+* Every Windows Terminal window is a "Peasant" process.
+* One of the Windows Terminal window processes is also the "Monarch" process.
+  The Monarch is picked randomly from the Terminal windows, and there is only
+  ever one Monarch process at a time.
+* Peasants can communicate with the monarch when certain state changes (such as
+  their window being activated), and the monarch can send commands to any of hte
+  peasants.
+
+This architecture will be used to enable each of the following scenarios.
+
+### Scenario: Open new tabs in most recently used window
+
+A common feature of many browsers is that when a web URL is clicked somewhere,
+the web page is opened as a new tab in the most recently used window of the
+browser. This functionality is often referred to as "glomming", as the new tab
+"gloms" onto the existing window.
+
+Currently, the terminal does not support such a feature - every `wt` invocation
+creates a new window. With the monarch/peasant architecture, it'll now be
+possible to enable such a scenario.
+
+As each window is activated, it will call a method on the `Monarch` object
+(hosted by the monarch process) which will indicate that "I am peasant N, and
+I've been focused". The monarch will use those method calls to update its own
+internal stack of the most recently used windows.
+
+Whenever a new `wt.exe` process is launched, that process will _first_ ask the
+monarch if it should run the commandline in an existing window, or create its
+own window.
+
+![auto-glom-wt-exe](auto-glom-wt-exe.png)
+
+If glomming is enabled, the monarch will dispatch the
+commandline to the appropriate window for them to handle instead. To the user,
+it'll seem as if the tab just opened in the most recent window.
+
+Users should certainly be able to specify if they want new instances to glom
+onto the MRU window or not. You could imagine that currently, we default to the
+hypothetical value `"glomToLastWindow": false`, meaning that each new wt gets
+it's own new window.
+
+If glomming is disabled, then the Monarch will call back to the peasant and tell
+it to run the provided commandline. The monarch will use the return value of
+`ExecuteCommandline` to indicate that the calling process should create a window
+and become a peasant process, and run the commandline itself.
+
+#### Glomming within the same virtual desktop
+
+When links are opened in the new Edge browser, they will only glom onto an
+existing window if that window is open in the current virtual desktop. This
+seems like a good idea of a feature for the Terminal to follow as well.
+
+There must be some way for an application to determine which virtual desktop it
+is open on. We could use that information to have the monarch track the last
+active window _per-desktop_, and only glom when there's one on the current
+desktop.
+
+We could make the `glomToLastWindow` property even more powerful by accepting a

```

Comment:
Would it be better to have two options. One is `newWindowOption`: `glom` or `newWindow`. The other is `glomOption`: `always`, `sameDesktop`.

Reply:
Ah but then, what would `"newWindowOption": "newWindow", "glomOption": "{anything}"` do? Presumably `glomOption` doesn't actually have a meaning in that scenario, right?

Reply:
With SUI in mind, ideally I'd like to disable `glomOption` if `newWindow` is on. But I guess we can not express this kind of dependency right now. Still I feel that two options are easier to understand than single option. If there's only just one option, we'll need more explanation in the doc , I guess.